### ISSUE TITLE
Long URLs causing SQL duplicate key errors

### ISSUE THREAD

**Issue by thelowlypeon** [nystudio107/retour#14]
tldr; Retour incorrectly attempts to insert a duplicate `Retour_StatsRecord` when the URLs are longer than 255 characters.

In `RetourService::incrementStatistics($url, $handled)`, Retour is looking for a record in `retour_stats` where `redirectSrcUrl =' . craft()->db->quoteValue($url)` ([here](https://github.com/nystudio107/retour/blob/master/services/RetourService.php#L320)). If one isn't found, it adds a new record with `redirectSrcUrl = $url` ([here](https://github.com/nystudio107/retour/blob/master/services/RetourService.php#L326)). 

In the former, if `$url` is more than 255 characters, no results will ever be returned. In the latter, the URL is being truncated in order to fit it into the table.

This means that if the URL is longer than 255 characters, Retour attempts to insert a new record into `retour_stats`, which has a unique key constraint on `redirectSrcUrl`, which fails because the truncated URL already exists in the DB.

Consider an example where the field is max 10 characters, where the actual url is `helloworld`:

1. Retour calls `incrementStatistics('/helloworld')`
2. `$result` is empty because no records exist where `redirectSrcUrl = '/helloworld'`
3. A new `Retour_StatsRecord` is created with `redirectSrcUrl = '/helloworld', which gets truncated on insert to `/helloworl`.
4. Another request is made for `/helloworld`, so repeat steps 1, 2, and 3. On step 3, the `INSERT` fails due to unique key `redirectSrcUrl = '/helloworl'`

**Comment by khalwat** [comment]
I have it truncating the incoming URLs for statistics to 255 characters for now; it's checked into `develop`:

https://github.com/nystudio107/retour/tree/develop

Probably a better solution long-term is that I create a migration to allow the URLs to be longer than 255 characters

**Comment by thelowlypeon** [comment]
thanks @khalwat! another option would be to do an `INSERT ... ON DUPLICATE KEY UPDATE ...` in a single query. might make that method simpler, but would move more of the logic out of the application and into sql. so, pros and cons.
